DM only when user breaks a step

thatportugueseguy opened the pull request:
With this change monorobot will only send one notification per failing/failed step, when a step is newly broken. If dev B's build has some failing steps from previous commit with dev A's code, dev B will not be notified. Only dev A will be notified, and only on the first time the build failed with that/those failing step(s). If there are new steps broken by dev C, while maintaining the original failed builds, dev C will only be notified about the steps they are responsible for breaking.

(Buildkite specific)
If the policy for failure is `allow_once`, there will be only one DM, sent after the whole build failed, and showing all the failed steps. Until these steps are fixed and broken again, there is no other notification for them.

If the policy for failure is `allow`, there will be two DMS with the above rules. The first will be on the notification for `build is failing`, the second will be on the notification for `build is failed`

thatportugueseguy commented on lib/action.ml:
```diff
@@ -128,71 +128,87 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
-          | Ok commit -> Lwt.return @@ partition_commit cfg commit.files)
-      in
-      Lwt.return (direct_message @ chans)
+      let broken_steps = Util.Build.new_failed_steps n repo_state pipeline in
+      match n.state = Failure && broken_steps = [] with
+      | true -> Lwt.return []
+      | false ->
```
this is the only change here, which changed the indentation of the `action_on_match` fn, hence the big diff, but the function itself wasn't changed

thatportugueseguy replied:
this is the only change, which changed the indentation of the `action_on_match` fn, hence the big diff, but the  function itself wasn't changed

thatportugueseguy commented on lib/action.ml:
```diff
@@ -128,71 +128,87 @@ module Action (Github_api : Api.Github) (Slack_api : Api.Slack) = struct
-          let notify_dm =
-            notify_dm && not (State.mem_repo_pipeline_commits ctx.state repo.url ~pipeline ~commit:n.sha)
+          let full_build_failed =
+            n.state = Failure && Re2.matches buildkite_is_failed_re (Option.default "" n.description)
           in
-          action_on_match branches ~notify_channels ~notify_dm
+          (* for failing states, if we only allow one notification, it must be the final one, not an intermediate one *)
+          (match n.state <> Failure || full_build_failed with
+          | false -> Lwt.return []
+          | true ->
```
same as above

thatportugueseguy replied:
same as above

thatportugueseguy commented on lib/common.ml:
```diff
@@ -40,32 +40,3 @@ module Re2 = struct
   let wrap s = create_exn s
   let unwrap = Re2.to_string
 end
-
-open Devkit
-
-let fmt_error ?exn fmt =
-  Printf.ksprintf
-    (fun s ->
-      match exn with
-      | Some exn -> Error (s ^ " :  exn " ^ Exn.str exn)
-      | None -> Error s)
-    fmt
-
-let first_line s =
-  match String.split_on_char '\n' s with
-  | x :: _ -> x
-  | [] -> s
-
-let decode_string_pad s = Stre.rstrip ~chars:"= \n\r\t" s |> Base64.decode_exn ~pad:false
-
-let http_request ?headers ?body meth path =
-  let setup h =
-    Curl.set_followlocation h true;
-    Curl.set_maxredirs h 1
-  in
-  match%lwt Web.http_request_lwt ~setup ~ua:"monorobot" ~verbose:true ?headers ?body meth path with
-  | `Ok s -> Lwt.return @@ Ok s
-  | `Error e -> Lwt.return @@ Error e
-
-let sign_string_sha256 ~key ~basestring =
-  Cstruct.of_string basestring |> Nocrypto.Hash.SHA256.hmac ~key:(Cstruct.of_string key) |> Hex.of_cstruct |> Hex.show
```
I've moved these more general/utility functions to the `util.ml` file as I wanted to add a function to this file, but it would cause cyclic dependencies. And these functions actually belong more in a util file than here, so I took the opportunity to move them

thatportugueseguy commented on test/monorobot.json:
This file only has formatting changes, apart from changing the policy for failure notifications from `allow_once` to `allow` so that we could still test the notifications for failing builds. Otherwise, we don't notify on failing builds, only on failed ones

thatportugueseguy commented on test/slack_payloads.expected:
i have updated the tests for `failure` status notifications and added a couple of extra ones. Cases tested:
- one failing step
- multiple failing steps at the same time
- someone breaks a new step while the previous steps keep failing

thatportugueseguy commented on mock_payloads/status.failure_test.json:
not sure this test case is really needed. Unless I'm missing something it tests the same thing as `status.commit1-02-failed`, but it was doing it before already, so I'm wondering if I might have overlooked something?

rr0gi commented:
ftr i want to know when my build is broken regardless whether it is my fault or not (ie exception should be configurable per user)

rr0gi commented:
maybe better to send dm to commit author anyway, just word it differently (build failed, but it is likely not your fault because build for step X was broken first in commit Y etc)

thatportugueseguy commented:
> also i think allow_once needs some renotification or escalation policy

I'm not sure I understand what you mean? Do you mean escalating to a message in a channel or DMing again if that step is broken for a long time?

thatportugueseguy commented:
> maybe better to send dm to commit author anyway, just word it differently (build failed, but it is likely not your fault because build for step X was broken first in commit Y etc)

I have updated the messages and added the link to the commit so that it's easy to see the code and the author, as well as the build logs. If it's justified, maybe we can add mentions later, but I wanted to move away from all that noise in the notification.

some screenshots:
<img width="489" alt="Screenshot 2024-10-02 at 18 53 35" src="https://github.com/user-attachments/assets/9898cf4c-ee06-479a-bdde-828622803a2b">
<img width="645" alt="Screenshot 2024-10-02 at 18 52 35" src="https://github.com/user-attachments/assets/451d97a7-654e-43bf-9b44-a8abd8f78dc6">
<img width="658" alt="Screenshot 2024-10-02 at 18 53 13" src="https://github.com/user-attachments/assets/c84408ac-0802-466f-a86c-844e4e5f92aa">

thatportugueseguy commented:
> also i think allow_once needs some renotification or escalation policy

maybe some kind of cron checking if each build has been failing for more than x mins and dm the original committer? I think this can/should be done one another PR

thatportugueseguy commented:
I'm going to merge this PR as I think it's fit to merge and we can look into the escalation mechanism after we test it.
